velox/dwio/parquet/reader/NestedStructureDecoder.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/dwio/parquet/reader/NestedStructureDecoder.h"
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/dwio/common/BufferUtil.h"
+
+namespace facebook::velox::parquet {
+
+int64_t NestedStructureDecoder::readOffsetsAndNulls(
+    const uint8_t* definitionLevels,
+    const uint8_t* repetitionLevels,
+    int64_t numValues,
+    uint8_t maxDefinition,
+    uint8_t maxRepeat,
+    BufferPtr& offsetsBuffer,
+    BufferPtr& lengthsBuffer,
+    BufferPtr& nullsBuffer,
+    memory::MemoryPool& pool) {
+  dwio::common::ensureCapacity<uint64_t>(
+      nullsBuffer, numValues / 64 + 1, &pool);
+  dwio::common::ensureCapacity<vector_size_t>(
+      offsetsBuffer, numValues + 1, &pool);
+  dwio::common::ensureCapacity<vector_size_t>(lengthsBuffer, numValues, &pool);
+
+  auto offsets = offsetsBuffer->asMutable<vector_size_t>();
+  auto lengths = lengthsBuffer->asMutable<vector_size_t>();
+  auto nulls = nullsBuffer->asMutable<uint64_t>();
+
+  int64_t lastCollectionEnd = 0;
+  bool wasLastCollectionNull = definitionLevels[0] == (maxDefinition - 1);
+  bits::setBit<uint64_t>(nulls, 0, wasLastCollectionNull);
+  offsets[0] = 0;
+
+  int64_t outputIndex = 1;
+  for (int64_t i = 1; i < numValues; ++i) {
+    uint8_t definitionLevel = definitionLevels[i];
+    uint8_t repetitionLevel = repetitionLevels[i];
```

Comment:
Isn't skipping the def/rep levels if they do not belong to this level more efficient here?

Reply:
@majetideepak That proved to be the opposite. Having branching is more detrimental to the perf. I tested 3 implementations, with two of them skipping non-collection-begin values, and they are much slower. My benchmark results:

No branching : 2.51ms
1 branching: 3.83ms
2 branchings: 4.89ms

1 Branching solution added the following at the beginning of the loop:
```
if (!isBegin) {
    lastOffset = isEmpty ? offset : lastCollectionEnd;
    continue;
}
```
2 branching solution added the following at the beginning of the loop:
```
if (repetitionLevel >= maxRepeat) {
  // Is not begin of a collection, do nothing.
  continue;
}

if (definitionLevel < (maxDefinition - 1)) {
  // The collection is empty. This marks the collection end.
  lastOffset = offset;
  continue;
}
```